src/DynamicData/Binding/SortAndBind.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) 2011-2023 Roland Pheasant. All rights reserved.
+// Roland Pheasant licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for full license information.
+
+using System.Reactive.Linq;
+using DynamicData.Cache;
+using DynamicData.Cache.Internal;
+
+namespace DynamicData.Binding;
+
+/*
+ * A much more optimised bind where the sort forms part of the binding.
+ *
+ * This is much efficient as the prior sort mechanism would resort and clone the entire
+ * collection upon every change in order that the sorted list could be transmitted to the bind operator.
+ *
+ */
+internal sealed class SortAndBind<TObject, TKey>(
+    IObservable<IChangeSet<TObject, TKey>> source,
+    IComparer<TObject> comparer,
+    SortAndBindOptions options,
+    IList<TObject> target)
+    where TObject : notnull
+    where TKey : notnull
+{
+    private readonly Cache<TObject, TKey> _cache = new();
+    private readonly object _locker = new();
+
+    public IObservable<IChangeSet<TObject, TKey>> Run() =>
+        source
+            .Synchronize(_locker)
+            // apply sorting as a side effect of the observable stream.
+            .Do(changes =>
+            {
+                // clone to local cache so that we can sort entire set when threshold is over a certain size.
+                _cache.Clone(changes);
+
+                // apply sorted changes to the target collection
+                if (target.Count == 0 || (options.ResetThreshold != 0 && options.ResetThreshold < changes.Count))
+                {
+                    Reset(_cache.Items.OrderBy(kv => kv, comparer));
+                }
+                else
+                {
+                    ApplyChanges(changes);
+                }
+            });
+
+    private void Reset(IEnumerable<TObject> sorted)
+    {
+        if (target is ObservableCollectionExtended<TObject> observableCollectionExtended)
+        {
+            using (observableCollectionExtended.SuspendNotifications())
+            {
+                observableCollectionExtended.Load(sorted);
+            }
+        }
+        else
+        {
+            target.Clear();
+            foreach (var t in sorted)
+            {
+                target.Add(t);
+            }
+        }
+    }
+
+    private void ApplyChanges(IChangeSet<TObject, TKey> changes)
+    {
+        // iterate through collection, find sorted position and apply changes
+
+        foreach (var change in changes.ToConcreteType())
+        {
+            var item = change.Current;
+
+            switch (change.Reason)
+            {
+                case ChangeReason.Add:
+                    {
+                        var index = GetInsertPosition(item);
+                        target.Insert(index, item);
+                    }
+                    break;
+                case ChangeReason.Update:
+                    {
+                        var currentIndex = GetCurrentPosition(change.Previous.Value);
```

Comment:
This might benefit a lot of cases to avoid copying all the items left to remove and then copying them right back to insert.  It is at the the expense of an extra check, but I think it is a good trade-off.

```csharp
                        var currentIndex = GetCurrentPosition(change.Previous.Value);
                        var index = GetInsertPosition(item);
                        
                        if (currentIndex != index)
                        { 
                            target.RemoveAt(currentIndex);
                            target.Insert(index, item);
                        }
                        else
                        { 
                            target[currentIndex] = item;
                        }
```

Reply:
Great point, I meant to account for this case but forgot.  However with anything to do with binding and sorting it is seriously complicated and there's a heap of hidden gotchas!

Some control suites and platforms do not support replace, whiles others do.  Remove and insert was the original behaviour in Dynamic Data and I later added the opt in option to use replace instead.  I'll include this change in this PR.

Reply:
Done